internal/xfn/function_runner.go

```diff
@@ -0,0 +1,283 @@
+// making a gRPC call to a Function package's runtime.
+func NewPackagedFunctionRunner(c client.Reader, o ...PackagedFunctionRunnerOption) *PackagedFunctionRunner {
+	r := &PackagedFunctionRunner{
+		client: c,
+		creds:  insecure.NewCredentials(),
+		conns:  make(map[string]*grpc.ClientConn),
+		log:    logging.NewNopLogger(),
+	}
+
+	for _, fn := range o {
+		fn(r)
+	}
+
+	return r
+}
+
+// RunFunction sends the supplied RunFunctionRequest to the named Function. The
+// function is expected to be an installed Function.pkg.crossplane.io package.
+func (r *PackagedFunctionRunner) RunFunction(ctx context.Context, name string, req *v1beta1.RunFunctionRequest) (*v1beta1.RunFunctionResponse, error) {
+	conn, err := r.getClientConn(ctx, name)
+	if err != nil {
+		return nil, errors.Wrapf(err, errFmtGetClientConn, name)
+	}
+
+	// This context is used for actually making the request.
+	ctx, cancel := context.WithTimeout(ctx, runFunctionTimeout)
+	defer cancel()
+
+	rsp, err := v1beta1.NewFunctionRunnerServiceClient(conn).RunFunction(ctx, req)
+	return rsp, errors.Wrapf(err, errFmtRunFunction, name)
+}
+
+// In most cases our gRPC target will be a Kubernetes Service. The package
+// manager creates this service for each active FunctionRevision, but the
+// Service is aligned with the Function. It's name is derived from the Function
+// (not the FunctionRevision). This means the target won't change just because a
+// new FunctionRevision was created.
+//
+// However, once the runtime config design is implemented it's possible that
+// something other than the package manager will reconcile FunctionRevisions.
+// There's no guarantee it will create a Service, or that the endpoint will
+// remain stable across FunctionRevisions.
+//
+// https://github.com/crossplane/crossplane/blob/226b81f/design/one-pager-package-runtime-config.md
+//
+// With this in mind, we attempt to:
+//
+// * Create a connection the first time someone runs a Function.
+// * Cache it so we don't pay the setup cost every time the Function is called.
+// * Verify that it has the correct target every time the Function is called.
+//
+// In the happy path, where a client already exists, this means we'll pay the
+// cost of listing and iterating over FunctionRevisions from cache. The default
+// RevisionHistoryLimit is 1, so for most Functions we'd expect there to be two
+// revisions in the cache (one active, and one previously active).
+func (r *PackagedFunctionRunner) getClientConn(ctx context.Context, name string) (*grpc.ClientConn, error) {
+	log := r.log.WithValues("function", name)
+
+	l := &pkgv1beta1.FunctionRevisionList{}
+	if err := r.client.List(ctx, l, client.MatchingLabels{pkgv1.LabelParentPackage: name}); err != nil {
+		return nil, errors.Wrapf(err, errListFunctionRevisions)
+	}
+
+	var active *pkgv1beta1.FunctionRevision
+	for i := range l.Items {
+		if l.Items[i].GetDesiredState() == pkgv1.PackageRevisionActive {
+			active = &l.Items[i]
```

Comment:
```suggestion
			active = &l.Items[i]
			break
```
Considering that there could only be one active revision.